File: paparazzi/src/main/java/com/squareup/paparazzi/internal/SessionParamsBuilder.java
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.squareup.paparazzi.internal;

import android.annotation.NonNull;
import com.android.SdkConstants;
import com.android.ide.common.rendering.api.AssetRepository;
import com.android.ide.common.rendering.api.LayoutlibCallback;
import com.android.ide.common.rendering.api.ResourceNamespace;
import com.android.ide.common.rendering.api.ResourceReference;
import com.android.ide.common.rendering.api.SessionParams;
import com.android.ide.common.rendering.api.SessionParams.RenderingMode;
import com.android.ide.common.resources.ResourceResolver;
import com.android.ide.common.resources.configuration.FolderConfiguration;
import com.android.ide.common.resources.deprecated.ResourceRepository;
import com.android.resources.ResourceType;
import com.google.common.collect.ImmutableMap;
import com.squareup.paparazzi.PaparazziLogger;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Builder to help setting up {@link SessionParams} objects.
 */
public class SessionParamsBuilder {

    private LayoutPullParser layoutPullParser;
    private RenderingMode renderingMode = RenderingMode.NORMAL;
    private Object projectKey = null;
    private DeviceConfig deviceConfig = DeviceConfig.NEXUS_5;
    private ResourceRepository frameworkResources;
    private ResourceRepository projectResources;
    private String themeName;
    private boolean isProjectTheme;
    private LayoutlibCallback layoutlibCallback;
    private int targetSdk;
    private int minSdk = 0;
    private PaparazziLogger logger;
    private Map<SessionParams.Key, Object> flags = new LinkedHashMap<>();
    private AssetRepository assetRepository = null;
    private boolean decor = true;

    @NonNull
    public SessionParamsBuilder setParser(@NonNull LayoutPullParser layoutParser) {
        this.layoutPullParser = layoutParser;

        return this;
    }

    @NonNull
    public SessionParamsBuilder setRenderingMode(@NonNull RenderingMode renderingMode) {
        this.renderingMode = renderingMode;
        return this;
    }

    @NonNull
    public SessionParamsBuilder setDeviceConfig(@NonNull DeviceConfig deviceConfig) {
        this.deviceConfig = deviceConfig;
        return this;
    }

    @NonNull
    public SessionParamsBuilder setProjectResources(@NonNull ResourceRepository resources) {
        projectResources = resources;
        return this;
    }

    @NonNull
    public SessionParamsBuilder setFrameworkResources(@NonNull ResourceRepository resources) {
        frameworkResources = resources;
        return this;
    }

    @NonNull
    public SessionParamsBuilder setTheme(@NonNull String themeName, boolean isProjectTheme) {
        this.themeName = themeName;
        this.isProjectTheme = isProjectTheme;
        return this;
    }

    @NonNull
    public SessionParamsBuilder setTheme(@NonNull String themeName) {
        boolean isProjectTheme;
        if (themeName.startsWith(SdkConstants.PREFIX_ANDROID)) {
            themeName = themeName.substring(SdkConstants.PREFIX_ANDROID.length());
            isProjectTheme = false;
        } else {
            isProjectTheme = true;
        }
        return setTheme(themeName, isProjectTheme);
    }

    @NonNull
    public SessionParamsBuilder setCallback(@NonNull LayoutlibCallback callback) {
        layoutlibCallback = callback;
        return this;
    }

    @NonNull
    public SessionParamsBuilder setTargetSdk(int targetSdk) {
        this.targetSdk = targetSdk;
        return this;
    }

    @SuppressWarnings("unused")
    @NonNull
    public SessionParamsBuilder setMinSdk(int minSdk) {
        this.minSdk = minSdk;
        return this;
    }

    @NonNull
    public SessionParamsBuilder setLogger(@NonNull PaparazziLogger logger) {
        this.logger = logger;
        return this;
    }

    @NonNull
    public SessionParamsBuilder setFlag(@NonNull SessionParams.Key flag, Object value) {
        flags.put(flag, value);
        return this;
    }

    @NonNull
    public SessionParamsBuilder setAssetRepository(@NonNull AssetRepository repository) {
        assetRepository = repository;
        return this;
    }

    @NonNull
    public SessionParamsBuilder disableDecoration() {
        decor = false;
        return this;
    }

    @NonNull
    public SessionParams build() {
        assert frameworkResources != null;
        assert projectResources != null;
        assert themeName != null;
        assert logger != null;
        assert layoutlibCallback != null;

        FolderConfiguration config = deviceConfig.getFolderConfig();
        ResourceResolver resourceResolver = ResourceResolver.create(
                ImmutableMap.of(
                        ResourceNamespace.ANDROID, frameworkResources.getConfiguredResources(config),
                        ResourceNamespace.TODO(), projectResources.getConfiguredResources(config)),
                new ResourceReference(
                        ResourceNamespace.fromBoolean(!isProjectTheme),
                        ResourceType.STYLE,
                    themeName));

        SessionParams params = new SessionParams(layoutPullParser, renderingMode, projectKey /* for
        caching */, deviceConfig.getHardwareConfig(), resourceResolver, layoutlibCallback,
            minSdk, targetSdk, logger);

        flags.forEach(params::setFlag);
        params.setAssetRepository(assetRepository);

        if (!decor) {
            params.setForceNoDecor();
        }

        return params;
    }
}

File: paparazzi/src/main/java/com/squareup/paparazzi/internal/SessionParamsBuilder.kt
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.squareup.paparazzi.internal

import com.android.SdkConstants
import com.android.ide.common.rendering.api.AssetRepository
import com.android.ide.common.rendering.api.LayoutlibCallback
import com.android.ide.common.rendering.api.ResourceNamespace
import com.android.ide.common.rendering.api.ResourceReference
import com.android.ide.common.rendering.api.SessionParams
import com.android.ide.common.rendering.api.SessionParams.Key
import com.android.ide.common.rendering.api.SessionParams.RenderingMode
import com.android.ide.common.resources.ResourceResolver
import com.android.ide.common.resources.ResourceValueMap
import com.android.ide.common.resources.deprecated.ResourceRepository
import com.android.resources.ResourceType
import com.squareup.paparazzi.PaparazziLogger
import java.util.LinkedHashMap

/**
 * Builder to help setting up [SessionParams] objects.
 */
class SessionParamsBuilder {
  private var layoutPullParser: LayoutPullParser? = null
  private var renderingMode = RenderingMode.NORMAL
  private val projectKey: Any? = null
  private var deviceConfig = DeviceConfig.NEXUS_5
  private var frameworkResources: ResourceRepository? = null
  private var projectResources: ResourceRepository? = null
  private var themeName: String? = null
  private var isProjectTheme: Boolean = false
  private var layoutlibCallback: LayoutlibCallback? = null
  private var targetSdk: Int = 0
  private var minSdk = 0
  private var logger: PaparazziLogger? = null
  private val flags = LinkedHashMap<SessionParams.Key<*>, Any>()
  private var assetRepository: AssetRepository? = null
  private var decor = true

  fun setParser(layoutParser: LayoutPullParser): SessionParamsBuilder {
    this.layoutPullParser = layoutParser
    return this
  }

  fun setRenderingMode(renderingMode: RenderingMode): SessionParamsBuilder {
    this.renderingMode = renderingMode
    return this
  }

  fun setDeviceConfig(deviceConfig: DeviceConfig): SessionParamsBuilder {
    this.deviceConfig = deviceConfig
    return this
  }

  fun setProjectResources(resources: ResourceRepository): SessionParamsBuilder {
    projectResources = resources
    return this
  }

  fun setFrameworkResources(resources: ResourceRepository): SessionParamsBuilder {
    frameworkResources = resources
    return this
  }

  fun setTheme(themeName: String, isProjectTheme: Boolean): SessionParamsBuilder {
    this.themeName = themeName
    this.isProjectTheme = isProjectTheme
    return this
  }

  fun setTheme(themeName: String): SessionParamsBuilder {
    var themeName = themeName
    val isProjectTheme: Boolean
    if (themeName.startsWith(SdkConstants.PREFIX_ANDROID)) {
      themeName = themeName.substring(SdkConstants.PREFIX_ANDROID.length)
      isProjectTheme = false
    } else {
      isProjectTheme = true
    }
    return setTheme(themeName, isProjectTheme)
  }

  fun setCallback(callback: LayoutlibCallback): SessionParamsBuilder {
    layoutlibCallback = callback
    return this
  }

  fun setTargetSdk(targetSdk: Int): SessionParamsBuilder {
    this.targetSdk = targetSdk
    return this
  }

  fun setMinSdk(minSdk: Int): SessionParamsBuilder {
    this.minSdk = minSdk
    return this
  }

  fun setLogger(logger: PaparazziLogger): SessionParamsBuilder {
    this.logger = logger
    return this
  }

  fun setFlag(flag: SessionParams.Key<*>, value: Any): SessionParamsBuilder {
    flags[flag] = value
    return this
  }

  fun setAssetRepository(repository: AssetRepository): SessionParamsBuilder {
    assetRepository = repository
    return this
  }

  fun disableDecoration(): SessionParamsBuilder {
    decor = false
    return this
  }

  fun build(): SessionParams {
    require(frameworkResources != null)
    require(projectResources != null)
    require(themeName != null)
    require(logger != null)
    require(layoutlibCallback != null)

    val folderConfiguration = deviceConfig.folderConfiguration
    val resourceResolver = ResourceResolver.create(
        mapOf<ResourceNamespace, Map<ResourceType, ResourceValueMap>>(
            ResourceNamespace.ANDROID to frameworkResources!!.getConfiguredResources(folderConfiguration),
            ResourceNamespace.TODO() to projectResources!!.getConfiguredResources(folderConfiguration)
        ),
        ResourceReference(
            ResourceNamespace.fromBoolean(!isProjectTheme),
            ResourceType.STYLE,
            themeName
        )
    )

    val result = SessionParams(layoutPullParser, renderingMode, projectKey /* for caching */,
        deviceConfig.hardwareConfig, resourceResolver, layoutlibCallback, minSdk, targetSdk, logger)

    for ((key, value) in flags) {
      result.setFlag(key as Key<Any>, value)
    }
    result.setAssetRepository(assetRepository)

    if (!decor) {
      result.setForceNoDecor()
    }

    return result
  }
}
